### ISSUE TITLE
Data race in node and network address book management

### ISSUE THREAD

**Issue by xin-hedera** [hashgraph/hedera-sdk-go#763]
### Description

Generally there lacks concurrent access protection in SDK's node and network address book management, which would cause a lot of data race problems especially when an application tries to submit high TPS traffic in many concurrent go routines.

### Steps to reproduce

refer to the description

### Hedera network

other

### Version

main

### Operating system

None

**Comment by rokn** [comment]
PR #765 includes the `-race` flag for unit tests, do you think that would be a sufficient coverage for now, or do you have some more specific issue?

**Comment by xin-hedera** [comment]
@rokn `-race` flag should be there for all `go test` in CI workflows and it actually makes more sense to do race detection in integration tests since concurrent writes / reads in node / address book management are more likely happen during integration tests.

**Comment by steven-sheehy** [comment]
The SDK race conditions are causing mirror node CI to frequently fail. Besides adding `-race` everywhere can the SDK team actively go through the address book management code and ensure it has proper locking in place?

**Comment by bamzedev** [comment]
There a some race conditions left in topic message query but they are coming in a second PR. This would be the last data race we are seeing when running the test with `-race`. If you believe there are more, feel free to reopen or create a new issue with more details. @xin-hedera

**Comment by xin-hedera** [comment]
I would disagree, @SimiHunjan?

though there’s no guarantee with -race flag all data races can be detected, we need to do a better job to fix the obvious ones.

**Comment by SimiHunjan** [comment]
let's reopen the issue

**Comment by SimiHunjan** [comment]
From Xin:

v2.27.0. I believe the address book update data races are addressed by the PRs mentioned above. However, large part if not the whole node management code still has data race. A lot of functions in managed_node.go need to be protected. The typical path when this happens:

- _Execute in executable.go
- pick a node and its channel to send the transaction
- check and update the node’s stats
- Anytime two or more go routines pick the same node to send a transaction, data races can happen

**Comment by rokn** [comment]
Hey @xin-hedera I've reinvestigated the issue with the help of your example, and I've discovered some issues we still had with the concurrency but with the merge of #808 the execute path should be clear of concurrency issues. Again as we've discussed before we won't be adding concurrency checks on some config level functions like some setters and getters as we don't want to over complicate the code. That shouldn't be a problem though as if a user needs those things to be thread safe, it can be done outside of the SDK. 

Thanks for the assist in discovering those issues!
